posts/kafka-tutorial-4-avro-and-schema-registry.md

```diff
@@ -0,0 +1,223 @@
+```
+
+# Producing Avro records
+
+Now, we want to change our producer code to send Avro data. The first thing to know is that there are two flavors of Avro records:
+- _specific records_: from the Avro schema, you generate Java classes using an Avro command - I don't like this approach too much though
+- _generic records_: you use a data structure that is pretty much like a map/dictionary, meaning you get/set the fields by their names and have to know their type. Although this is not type-safe, this offers a lot more flexibility and is much easier to work with when your schema changes over time - this is the approach we are going to use here.
+
+Before we can start coding, we need to add the library that adds Avro support to the Kafka client - this library is stored in Confluent's Maven repository:
+
+```gradle
+repositories {
+    mavenCentral()
+    maven { url 'https://packages.confluent.io/maven/' }
+}
+dependencies {
+    ...
+    compile 'io.confluent:kafka-avro-serializer:5.0.0'
+}
+```
+
+Let's go ahead and modify our producer:
+
+```kotlin
+private fun createProducer(brokers: String, schemaRegistryUrl: String): Producer<String, GenericRecord> {
+    val props = Properties()
+    props["bootstrap.servers"] = brokers
+    props["key.serializer"] = StringSerializer::class.java
+    props["value.serializer"] = KafkaAvroSerializer::class.java
+    props["schema.registry.url"] = schemaRegistryUrl
+    return KafkaProducer<String, GenericRecord>(props)
+}
+```
+
+Three things to note:
+1. We switched to using the `KafkaAvroSerializer` for the value only, and kept the key serializer as is.
+2. We added the URL of the Schema Registry under the `schema.registry.url` property.
+3. We changed the type of the producer to accept objects of type `GenericRecord` for the value.
+
+Now, because we are going to use generic records, we need to load the schema. We could get it from the Schema Registry, but this is not very safe because you don't know what version of the schema has been registered. Instead, **it is a good practice to store the schema alongside the code**. That way, your code always produces the right type of data, even if someone else changes the schema registered in the Schema Registry.
+
+```kotlin
+val schema = Schema.Parser().parse(File("src/main/resources/person.avsc"))
+```
+
+We can now create `GenericRecord` objects using a `GenericRecordBuilder`:
+
+```kotlin
+val avroPerson = GenericRecordBuilder(schema).apply {
+    set("firstName", fakePerson.firstName)
+    set("lastName", fakePerson.lastName)
+    set("birthDate", fakePerson.birthDate.time)
+}.build()
+```
+
+As in the first post, we used Kotlin's `apply` method to avoid repeating code.
+
+We can now send the record:
+
+```kotlin
+val futureResult = producer.send(ProducerRecord(personsAvroTopic, avroPerson))
+```
+
+# Testing the code
```

Comment:
It would be nice if this section contained an automated test :)

Reply:
Yes, that's something I have in mind to talk about in a further post.